Source/Particles/Collision/BackgroundMCCCollision.cpp

```diff
@@ -0,0 +1,373 @@
+        // if the scattering process is excitation or ionization get the
+        // energy associated with that process
+        if (scattering_process.find("excitation") != std::string::npos ||
+            scattering_process.find("ionization") != std::string::npos)
+        {
+            std::string temp = scattering_process;
+            temp.append("_energy");
+            char kw[temp.length() + 1];
+            std::strcpy(kw, temp.c_str());
+            pp.get(kw, energy);
+        }
+
+        // if the scattering process is ionization get the secondary species
+        // only one ionization process is supported, the vector
+        // m_ionization_processes is only used to make it simple to calculate
+        // the maximum collision frequency with the same function used for
+        // particle conserving processes
+        if (scattering_process == "ionization")
+        {
+            ionization_flag = true;
+
+            std::string secondary_species;
+            pp.get("ionization_species", secondary_species);
+            m_species_names.push_back(secondary_species);
+
+            m_ionization_processes.push_back(
+                MCCProcess(scattering_process, cross_section_file, energy)
+            );
+        }
+        else
+        {
+            m_scattering_processes.push_back(
+                MCCProcess(scattering_process, cross_section_file, energy)
+            );
+        }
+    }
+}
+
+/** Calculate the maximum collision frequency using a fixed energy grid that
+ *  ranges from 1e-4 to 5000 eV in 0.2 eV increments
+ */
+amrex::Real
+BackgroundMCCCollision::get_nu_max(amrex::Vector<MCCProcess> mcc_processes)
+{
+    amrex::Real nu, nu_max = 0.0;
+
+    for (double E = 1e-4; E < 5000; E+=0.2)
+    {
+        amrex::Real sigma_E = 0.0;
+
+        // loop through all collision pathways
+        for (auto scattering_process : mcc_processes){
+            // get collision cross-section
+            sigma_E += scattering_process.getCrossSection(E);
+        }
+
+        // calculate collision frequency
+        nu = (
+            m_background_density * std::sqrt(2.0 / m_mass1 * PhysConst::q_e)
+            * sigma_E * std::sqrt(E)
+        );
+        if (nu > nu_max)
+        {
+            nu_max = nu;
+        }
+    }
+    return nu_max;
+}
+
+void
+BackgroundMCCCollision::doCollisions (amrex::Real cur_time, MultiParticleContainer* mypc)
+{
+    WARPX_PROFILE("BackgroundMCCCollision::doCollisions()");
+
+    const amrex::Real dt = WarpX::GetInstance().getdt(0);
+    if ( int(std::floor(cur_time/dt)) % m_ndt != 0 ) return;
+
+    auto& species1 = mypc->GetParticleContainerFromName(m_species_names[0]);
+    // this is a very ugly hack to have species2 be a reference and be
+    // defined in the scope of doCollisions
+    auto& species2 = (
+        (m_species_names.size() == 2) ?
+            mypc->GetParticleContainerFromName(m_species_names[1]) :
+        mypc->GetParticleContainerFromName(m_species_names[0])
```

Comment:
I'm curious if people have ideas for a better way to do this.

Reply:
It's a bit ugly, but it looks reasonable enough to me. I don't think it's going to hurt anything.